src/mnode/mnode-pastel.cpp

```diff
@@ -1571,3 +1571,175 @@ bool CTakeDownTicket::FindTicketInDb(const std::string& key, CTakeDownTicket& ti
+    
+    std::string strTicket = ticket.ToStr();
+    ticket.signature = CPastelID::Sign(reinterpret_cast<const unsigned char*>(strTicket.c_str()), strTicket.size(), ticket.pastelID, strKeyPass);
+
+    return ticket;
+}
+
+bool CChangeUsernameTicket::FindTicketInDb(const std::string& key, CChangeUsernameTicket& ticket)
+{
+    ticket.username = key;
+    return masterNodeCtrl.masternodeTickets.FindTicket(ticket);
+}
+
+std::string CChangeUsernameTicket::ToJSON() const noexcept
+{
+    json jsonObj;
+    jsonObj = {
+        {"txid", m_txid},
+        {"height", m_nBlock},
+        {"ticket", {
+            {"type", GetTicketName()}, 
+            {"pastelID", pastelID},
+            {"username", username}, 
+            {"fee", fee}, 
+            {"signature", ed_crypto::Hex_Encode(signature.data(), signature.size())}}}};
+
+    return jsonObj.dump(4);
+}
+
+std::string CChangeUsernameTicket::ToStr() const noexcept
+{
+    std::stringstream ss;
+    ss << pastelID;
+    ss << username;
+    ss << fee;
+    ss << m_nTimestamp;
+    return ss.str();
+}
+
+// FIXME: Remove erret parameters
+bool CChangeUsernameTicket::IsValid(bool preReg, int depth) const
+{
+    unsigned int chainHeight = 0;
+    {
+        LOCK(cs_main);
+        chainHeight = static_cast<unsigned int>(chainActive.Height()) + 1;
+    }
+
+    // A. Something to check ONLY before ticket made into transaction
+    if (preReg) 
+    {
+        // A1. Check if the username is already registered on the blockchain.
+        if (masterNodeCtrl.masternodeTickets.CheckTicketExist(*this)) {
+            throw std::runtime_error(strprintf("This Username is already registered in blockchain [Username = %s]", username));
+        }
+        // A2. Check if address has coins to pay for Username Change Ticket
+        const auto fullTicketPrice = TicketPrice(chainHeight);
+
+        if (pwalletMain->GetBalance() < fullTicketPrice * COIN) {
+            throw std::runtime_error(strprintf("Not enough coins to cover price [%" PRId64 "]", fullTicketPrice));
+        }
+    }
+
+    // Check if username is a bad username. For now check if it is empty only.
+    std::string badUsernameError;
+    if (isUsernameBad(username, badUsernameError)) {
+        throw std::runtime_error(strprintf(badUsernameError.c_str()));
+    }
+
+    // B Verify signature
+    // We will check that it is the correct PastelID
+    std::string strThisTicket = ToStr();
+    if (!CPastelID::Verify(reinterpret_cast<const unsigned char *>(strThisTicket.c_str()), strThisTicket.size(),
+                           signature.data(), signature.size(),
+                           pastelID))
+    {
+        throw std::runtime_error(strprintf("%s ticket's signature is invalid. PastelID - [%s]", GetTicketDescription(TicketID::Username), pastelID));
+    }
+    
+    // C (ticket transaction replay attack protection)
+    CChangeUsernameTicket existingTicket;
+    if ((FindTicketInDb(username, existingTicket)) &&
```

Comment:
it does not duplicate CheckTicketExist above? How it will work if user1 changed the 'name1' to 'name2' and another user2 wants to set his username to 'name1'?

Reply:
it is allowed as discussed with the team.

Reply:
I'm just worry we have 2 repeated calls to database that have the same result.

Reply:
they are used in different context, different conditions. Please check if `(pReg)`,  and also the conditions of `FindTicketInDb()`

Reply:
Probably it will be throw on CheckTicketExist and will not come here. I think you need the first to test next: "when user1 changed the 'name1' to 'name2' and another user2 wants to set his username to 'name1'". Probably you will need to do more updates, I did not think about like test case yet. I can think about this if you need help.

Reply:
that test case is tested already, not sure why it is removed in mn_tickets.py. Will add back.